.github/workflows/mkosi.yml

```diff
@@ -57,24 +57,33 @@ jobs:
           - distro: arch
             release: rolling
             sanitizers: ""
+            llvm: 0
           - distro: debian
             release: testing
             sanitizers: ""
+            llvm: 0
           - distro: ubuntu
             release: noble
             sanitizers: ""
+            llvm: 0
           - distro: fedora
             release: "40"
             sanitizers: ""
+            llvm: 0
           - distro: fedora
             release: rawhide
             sanitizers: address,undefined
+            llvm: 1
```

Comment:
As far as I understand systemd compiles with `-O0` by default here. https://clang.llvm.org/docs/AddressSanitizer.html#usage recommends adding -O1 or higher to get a reasonable performance.

Reply:
We build with Og in CI, which is like O1 except optimizations that interfere with debugging are disabled.

Reply:
I'm not sure how `-Og` helps with ASan/UBSan CI builds. Other than that I think CentOS CI already compiles systemd with clang+ASan/UBSan/-Og so in terms of covering more build profiles `-g -O1` or `-g -O2` would be more helpful.

Reply:
Looking at https://github.com/systemd/systemd-centos-ci/pull/200 @mrc0mmand I wonder if you remember what was stripped in ASan builds? Would `-fno-optimize-sibling-calls` have helped there?

Reply:
I'm guessing it improved the coredump collector but I don't think mkosi collects coredumps currently. I'm not sure if it's planned either.

Reply:
@evverx mkosi configures systemd-coredump to store all coredumps in the journal so we do collect them in the same way that the bash runner collects them.

Reply:
CentOS CI goes over those coredumps, extracts the relevant backtraces with gdb and shows them. It also fails if the testsuite itself passes but unexpected coredumps are detected. It's separate from the bash runner.

Reply:
Speaking of backtraces I wonder if the mkosi CI collects ASan/UBsan backtraces? I looked for something like
```
        pids="$(
            "$JOURNALCTL" -D "$root/var/log/journal" --grep 'SUMMARY: .*Sanitizer:' |
                   grep -v -E 'dbus-daemon|dbus-broker-launch' |
                   sed -r -n 's/.* .+\[([0-9]+)\]: SUMMARY:.*/\1/p'
        )"
```
in the code related to the mkosi but couldn't find anything like that. It could be I screwed something up though.

Reply:
@evverx I just pushed a commit that makes address sanitizer failures generate coredumps. So this will be covered automatically once I add a bit of logic to check if there are unexpected coredumps after running a test.

Reply:
> I don't think we need any further post processing of coredumps beyond checking if there are any. Developers can just download the journal, run coredumpctl --file <journal> and they'll get a list of coredumps.

Dunno. I always go the CentOS logs exactly because it automates all those annoying things. Ideally it shouldn't be necessary to manually download journals and go over them. Backtraces shown by llvm-symbolizer are usually much more useful than whatever is in coredumps.

Reply:
Those coredumps are huge and unlikely to be saved. CentOS CI ran into that https://github.com/systemd/systemd-centos-ci/issues/465#issuecomment-1050724484. FWIW I'd wait for @mrc0mmand here because I think CentOS CI ran into everything so some things can be borrowed from there I think.

Reply:
@evverx The huge coredump problem is fixed by enabling the `use_madv_dontdump` option. That makes sure the coredumps are of regular size.

Reply:
I ran it locally and got https://github.com/systemd/systemd/pull/33146#issuecomment-2143806459
```
Process 232841 (systemd-resolve) of user 524288 terminated abnormally without generating a coredump
```

Reply:
I'm not sure it's possible to collect coredumps on GitHub actions in general. Last time I checked they all went to the underlying host. It would probably make sense to add UB and a crash to a systemd service and find out whether coredumps are stored anywhere.

Reply:
@evverx Coredumps in containers are a bit of a problem yes, though the situation will improve once we switch to Noble since then we'll have #29272

Reply:
I don't think it can be controlled by GitHub Actions. The underlaying host should allow that too.

Either way I think it would be great if ASan coredumps were configurable because I don't think they are helpful. I already have 9944 "coredump"s in the container (fortunately with no corefiles thanks to the resource limits).

Reply:
@evverx I mean that when we switch to Noble as the host distribution on Github Actions, coredumps will be automatically forwarded to nspawn containers and end up in the container instead of on the host.

> Either way I think it would be great if ASan coredumps were configurable because I don't think they are helpful. I already have 9944 "coredump"s in the container (fortunately with no corefiles thanks to the resource limits).

mkosi supports dropins so you can add mkosi.conf.d/30-sanitizers-evverx.conf and override the asan options from 20-sanitizers.conf in whatever way you'd like.

Reply:
> mkosi supports dropins so you can add mkosi.conf.d/30-sanitizers-evverx.conf and override the asan options from 20-sanitizers.conf in whatever way you'd like

I get that. What I'm getting at is that in most cases coredumps aren't needed because backtraces are enough and better. I understand that to make it easier to detect failures on GitHub Actions it kind of makes sense to turn ASan failures into coredumps but it's likely to be the only place where this is needed.

Reply:
Looking at
```
Sun 2024-06-02 10:55:45 UTC 232841 524288 524288 SIGABRT none     /work/build/systemd-resolved         -
Sun 2024-06-02 11:26:06 UTC 254633 524288 524288 SIGABRT none     /work/build/systemd-resolved         -
Sun 2024-06-02 12:03:23 UTC 277594 525282 525282 SIGABRT none     /usr/lib/systemd/systemd-coredump    -
Sun 2024-06-02 12:03:32 UTC 277873 524481 524481 SIGABRT none     /usr/lib/systemd/systemd-resolved    -
Sun 2024-06-02 12:09:49 UTC 288779 525282 525282 SIGABRT none     /usr/lib/systemd/systemd-coredump    -
```
on the host it seems some coredumps (/usr/lib/*) weren't forwarded to the container so to make it robust the host should be checked anyway.

Reply:
@evverx Can you give some more information on the environment you're running this in?

At least on Fedora 40, if I kill systemd-resolved in the `mkosi boot` nspawn container, I get the following when running `coredumpctl` in the container afterwards:

```
[root@system ~]# coredumpctl
TIME                         PID UID GID SIG     COREFILE EXE                               SIZE
Sun 2024-06-02 17:57:28 CEST  88 193 193 SIGABRT journal  /usr/lib/systemd/systemd-resolved    -
```

Reply:
I built a Fedora image with clang/ASan/UBSan on a Fedora 40 host and introduced UB to systemd-resolved to see how it all works together. I got the "/work/build/systemd-resolved" coredumps on the host when the unit tests were run (because it happened when the image was built). Then I rebuilt the image with `-T` to skip the unit tests and booted the container with `mkosi` boot. When I ran `systemd-resolved` it failed under UBSan and generated a lot of coredumps thanks to autorestart (and some of them were forwarded to the container). `systemd-coredump` itself failed in the container with
```
Jun 02 11:53:24 systemd-coredump[123]: Processing coredump received on stdin...
Jun 02 11:53:24 systemd-coredump[123]: Resource limits disable core dumping for process 115 (systemd-coredum).
Jun 02 11:53:24 systemd-coredump[123]: [🡕] Process 115 (systemd-coredum) of user 994 terminated abnormally without generating a coredump.
Jun 02 11:53:24 systemd-coredump[123]: ==123==LeakSanitizer has encountered a fatal error.
Jun 02 11:53:24 systemd-coredump[123]: ==123==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
Jun 02 11:53:24 systemd-coredump[123]: ==123==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```
In the end I had 9944 coredumps in the container and a bunch of coredumps on the host that should have been forwarded to the container.

Reply:
@evverx The reason that there are some coredumps on the host is because systemd-coredump will fall back to the old behavior if systemd-coredump in the container for some reason fails to process the coredump. So if we figure out the leak sanitizer error that will automatically resolve itself as well.

Reply:
I can't seem to reproduce either error though, leak sanitizer never triggers for systemd-coredump in my own testing after introducing undefined behavior in systemd-resolved. All coredumps are correctly forwarded to the container.

Reply:
systemd-coredump fails under LSan because it drops its privileges at some point. It can be fixed either by providing a way to turn it off when it's run under LSan or by adding `detect_leaks=0` to `ASAN_OPTIONS` of `systemd-coredump@.service`.

`mkosi` still produces coredumps on the host when the image is built and the unit tests fail so coredumps/journals from containers aren't enough to detect all failures. I really think that the discussion on how mkosi should handle test failures, ASan failures, unexpected coredumps and so on should be moved to a separate issue where that part can be designed properly. I think that ASan/UBSan backtraces should be handled separately and shown when the tests fail to avoid downloading journals manually and manually looking for backtraces. I don't think mkosi CI should be another Ubuntu CI with its UX from 1998 after all.

Reply:
@evverx Feel free to open another issue for the coredump handling, it doesn't really relate to this PR in the first place. I will drop the `abort_on_error` commit from this PR.

Reply:
@evverx I just pushed a commit that allows using `mkosi -E MESON_VERBOSE=1` to enable verbose output from meson. I'm not yet enabling this in CI as this is insanely verbose and I don't think it makes sense to enable it by default without some kind of output grouping so we don't show it by default in CI logs.

Reply:
I'm testing it with FORTIFY_SOURCE now. Just out of curiosity is it expected that several CFLAGS should be double-quoted in the sense that they should be passed via
```
-E CFLAGS='"-D_FORTIFY_SOURCE=3 -O2"'
```
? Without the double quotes only `-D_FORTIFY_SOURCE=3` makes it into the build.

Reply:
Yes that's expected otherwise we parse it as multiple environment variables. This is a quirk of mkosi's parser.